google/cloud/storage/internal/async/reader_connection_resume_test.cc

```diff
@@ -0,0 +1,386 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/async/reader_connection_resume.h"
+#include "google/cloud/storage/async/resume_policy.h"
+#include "google/cloud/storage/internal/async/reader_connection_impl.h"
+#include "google/cloud/storage/internal/grpc/ctype_cord_workaround.h"
+#include "google/cloud/storage/mocks/mock_async_reader_connection.h"
+#include "google/cloud/storage/options.h"
+#include "google/cloud/storage/retry_policy.h"
+#include "google/cloud/storage/testing/canonical_errors.h"
+#include "google/cloud/testing_util/mock_async_streaming_read_rpc.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include <google/protobuf/text_format.h>
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using ::google::cloud::storage::testing::canonical_errors::TransientError;
+using ::google::cloud::storage_experimental::ReadPayload;
+using ::google::cloud::storage_experimental::ResumePolicy;
+using ::google::cloud::testing_util::IsOk;
+using ::testing::_;
+using ::testing::AllOf;
+using ::testing::AtMost;
+using ::testing::ElementsAre;
+using ::testing::Eq;
+using ::testing::IsEmpty;
+using ::testing::Pair;
+using ::testing::ResultOf;
+using ::testing::Return;
+using ::testing::UnorderedElementsAre;
+using ::testing::VariantWith;
+
+using MockReader = ::google::cloud::storage_mocks::MockAsyncReaderConnection;
+using ReadResponse =
+    ::google::cloud::storage_experimental::AsyncReaderConnection::ReadResponse;
+
+using MockAsyncReaderConnectionFactory = ::testing::MockFunction<future<
+    StatusOr<std::unique_ptr<storage_experimental::AsyncReaderConnection>>>(
+    storage::Generation, std::int64_t)>;
+
+class MockResumePolicy : public ResumePolicy {
+ public:
+  ~MockResumePolicy() override = default;
+  MOCK_METHOD(void, OnStartSuccess, (), (override));
+  MOCK_METHOD(ResumePolicy::Action, OnFinish, (Status const&), (override));
+};
+
+auto WithGeneration(std::int64_t expected) {
+  return ::testing::ResultOf(
+      "generation value is",
+      [](storage::Generation const& g) { return g.value_or(0); },
+      ::testing::Eq(expected));
+}
+
+auto WithoutGeneration() {
+  return ::testing::ResultOf(
+      "generation is not set",
+      [](storage::Generation const& g) { return g.has_value(); },
+      ::testing::Eq(false));
+}
+
+auto constexpr kReadSize = 500;
+auto constexpr kRangeStart = 10000;
+
+auto MakeMockReaderPartial(int offset)
```

Comment:
`std::int64_t` instead of `int`?

Reply:
Done.  Maybe I will learn something: why do you think that is better?

Reply:
Only that it matches the type of the object where it will be used.  You could, for example, want to make a (mock) reader with an offset beyond the range of an `int`, no?